src/components/ContentUploader/ContentUploader.js

```diff
@@ -233,57 +236,64 @@ class ContentUploader extends Component<Props, State> {
+        return Array.from(files).filter((file) => !(getFileId(file, rootFolderId) in itemIds));
     };
 
     /**
-     * Generates file id based on file properties
-     *
-     * When folderId or uploadInitTimestamp is missing from file options, file name is returned as file id.
-     * Otherwise, fileName_folderId_uploadInitTimestamp is used as file id.
+     * Given an array of files, return the files that are new to the Content Uploader
      *
-     * @param {UploadFileWithAPIOptions | UploadFile} file
+     * @param {Array<UploadFileWithAPIOptions | File>} files
      */
-    getFileId(file: UploadFileWithAPIOptions | UploadFile) {
+    getNewDataTransferItems = (items: Array<DataTransferItem | UploadDataTransferItemWithAPIOptions>) => {
         const { rootFolderId } = this.props;
+        const { itemIds } = this.state;
 
-        if (!doesFileContainAPIOptions(file)) {
-            return ((file: any): UploadFile).name;
-        }
-
-        const fileWithOptions = ((file: any): UploadFileWithAPIOptions);
-        const folderId = getProp(fileWithOptions, 'options.folderId', rootFolderId);
-        const uploadInitTimestamp = getProp(fileWithOptions, 'options.uploadInitTimestamp', 0);
-        const fileName = fileWithOptions.file.webkitRelativePath || fileWithOptions.file.name;
-
-        return `${fileName}_${folderId}_${uploadInitTimestamp}`;
-    }
+        return Array.from(items).filter((item) => !(getDataTransferItemId(item, rootFolderId) in itemIds));
+    };
 
     /**
      * Converts File API to upload items and adds to upload queue.
      *
      * @private
      * @param {Array<UploadFileWithAPIOptions | UploadFile>} files - Files to be added to upload queue
      * @param {Function} itemUpdateCallback - function to be invoked after items status are updated
+     * @param {boolean} [isRelativePathIgnored] - if true webkitRelativePath property is ignored
      * @return {void}
      */
-    addFilesToUploadQueue = (files?: Array<UploadFileWithAPIOptions | UploadFile>, itemUpdateCallback: Function) => {
+    addFilesToUploadQueue = (
+        files?: Array<UploadFileWithAPIOptions | UploadFile>,
+        itemUpdateCallback: Function,
+        isRelativePathIgnored?: boolean = false
+    ) => {
+        const { rootFolderId } = this.props;
         if (!files || files.length === 0) {
             return;
         }
 
+        const { itemIds } = this.state;
+        const newFiles = this.getNewFiles(files);
+
+        if (newFiles.length === 0) {
+            return;
+        }
+
+        newFiles.forEach((file) => {
+            itemIds[getFileId(file, rootFolderId)] = true;
+        });
+
         clearTimeout(this.resetItemsTimeout);
 
-        const firstFile = getFile(files[0]);
+        const firstFile = getFile(newFiles[0]);
 
-        if (firstFile.webkitRelativePath) {
-            this.addFilesWithRelativePathToQueue(files, itemUpdateCallback);
+        if (firstFile.webkitRelativePath && !isRelativePathIgnored) {
```

Comment:
Can you add some comments here? It's not clear to me why we only add the first item in this situation

Reply:
Thanks for you offline feedback, I updated the FolderUpload class interface which now only represents one folder.